Refactor into separate files, additional features, addition of missing typing

michalpokusa opened the pull request:
This PR, other that separating adafruit_httpserver into multiple files #8  adds following functionality:

-  Access to additional attributes of `HTTPRequest`:
    - **`http_version`** extracted from Start line, e.g. `"HTTP/1.1"`
    - **`query_params`** extracted from `path`, e.g. _`/path?foo=bar`_ is parsed to `{"foo": "bar"}`
    - **`headers`** as a `dict`, e.g. `{'User-Agent': '...', 'Host': 'esp32-s2-tft.local', 'Connection': 'close'}`
    - **`body`** as `bytes`, so both text content and images can be processed #16
- Added new parameter to `HTTPResponse` which allows to set it's `headers` #24 

Other thant that, this PR also includes:
- Refactor or addition of "enums" like `HTTPMethod` and `MIMEType`
- Removed blocking socket, wich fixes #21
- Smaller refactor here and there
- More typing in functions/classes etc.

Despite major refactor, the usage of library stays nearly unchanged, so it won't be neccessary to rewrite existing projects from scratch, only minor changes might me required.
It surely is not perfect, but I belive it is a step in the right direction.

Closes #8 #16 #17 #21 #24

dhalbert commented:
Take a look at using `pre-commit` if you want to test the black and pylint errors locally: https://learn.adafruit.com/creating-and-sharing-a-circuitpython-library/check-your-code

anecdata commented:
I'm not able to get the `examples/httpserver_simplepolling.py` to work from a browser. I put some debug prints in and it looks like the socket is accepted, but the `received_data` is sometimes empty, but even when it looks complete the server just hangs and never returns the response. The desktop browser just keeps waiting, I think browsers tend to have infinite timeout. I'll keep at it.

anecdata reviewed (commented):
A couple of questions. The only way I could complete a request from a browser was to use an arbitrary timeout, incurring a performance hit and assuming that whatever was received in that time was complete.

anecdata commented on adafruit_httpserver/server.py:
```diff
@@ -0,0 +1,200 @@
+                while "Receiving data":
+                    try:
+                        length = conn.recv_into(self._buffer)
+                        received_data += self._buffer[:length]
```
In the blocking case, we need a way to know when the request is complete?

anecdata replied:
Realized there is no blocking case currently allowed (timeout must be >=0, not `None`), but perhaps there should be?

michalpokusa replied:
I agree that giving an option to set timeout to `None` (blocking) should be possible, thus not liming if one wants to do so.

It might also be logical to increase default timeout to `0.5` or `1` that would allow receiving request without explicitly setting timeout using `server.socket_timeout`.

michalpokusa replied:
To be honest I do not really see any other way of determining whether it is the end of mesage other than timeout, I do not like that it unnecessarily waits after last part of message is received but it is the only way that seems to work no matter which client wants to connect to server. But, compared to current `main` branch, each request takes at least 2x longer.

One positive here is that it makes it possible to send data longer that max buffer size of 2880 bytes, which makes it more universal.

It is hard to find a middle point here, and we have to remember that we are talking about microcontrollers, not load-balanced web servers, so maybe a longer delay is not as much of a problem as the buffer size limit.

anecdata commented on adafruit_httpserver/server.py:
```diff
@@ -0,0 +1,200 @@
+                    try:
+                        length = conn.recv_into(self._buffer)
+                        received_data += self._buffer[:length]
+                    except OSError as ex:
```
In the non-blocking (no timeout) case, we need to handle EAGAIN here?

michalpokusa commented:
After more testing, it seems to me that the most reliable way of solving the socket blocking problem is making it possible **only to set it to positive number**, no `0` and `None`, this way eliminates infinitely blocked socket and enables request exceeding 2880 bytes.

**I know it is not perfect** and has an impact on loading speed, but I don't see any other way around it. Even if library api enabled setting `socket_timeout` to `0` or `None`, then it would simple hang on forever on first request. Preventing that from happening I believe is more user friendly.

Maybe there is other solution that I am missing, I do not completely understand all the quirks of TCP sockets and blocking behaviour. But from testing this way seems most reasonable to me and the impact itself is not that noticeable.

dhalbert commented:
What is different about the way HTTPServer is operating compared to, say, the simple CPython HTTP server, or Apache, nginx, etc. Do we need to fix something underlying?

michalpokusa commented:
As I mentioned above, I am not very familiar with TCP communication on low-level, but I found that a client should send a `shutdown(SHUT_WR)` to `socket.recv()` (which is not implemented yet in CircuitPython) to indicate the end of message [[Detect client is done with sending (shutdown)](https://iximiuz.com/en/posts/writing-web-server-in-python-sockets/#:~:text=Detect%20client%20is%20done%20with%20sending%20(shutdown))].

Also when analysing CPython and going deeper in call stack i found similar condition. [[CPython/Lib/socketserver.py:216](https://github.com/python/cpython/blob/56a517e9e854559e2f0c23c0e5adcaf846a2e0c3/Lib/socketserver.py#L232)]

Maybe that is the problem, but it is hard for me to say, as it is very low-level.

anecdata commented:
The issue with any timeout, aside from just the dead time, is that it can trip up async by extending the minimum overall async loop time.

Someone do please correct me if I'm wrong, but my understanding is that HTTP/1.1 adds some protocol assumptions over simple TCP: a request will either have a `transfer-encoding (e.g., chunked) OR a `content-length` header, or (to maintain compatibility with HTTP/1.0 clients) will handle if the client sends neither header but closes the socket - but that's strongly discouraged in HTTP/1.1. Does that not trigger a `ECONNRESET`? If not we can still have a timeout applied on top of handling `content-length`, and return `NotImplemented` for for now for `transfer-encoding`?

michalpokusa commented:
Finally I was able to get the expected result. As @anecdata suggested I used `Content-Length` to determine when there is no more data expected. Now timeout is **not** the only thing that can end the connection and it works with curl, browsers, adafruit_requests and other clients I tested, and is is fast, much faster than before.

I adjusted default `socket_timeout` and prevented from setting it to `None` or `0`.

I also changed how `HTTPRequest.headers` are parsed, as they, unlike query params, are case-insensitive. I used the same approach as implemented in [adafruit_wsgi/request.py:103](https://github.com/adafruit/Adafruit_CircuitPython_WSGI/blob/main/adafruit_wsgi/request.py#L103), which means that the are all lower case.

Right now there is no support for `Transfer-Encoding`, but I believe it is better to make it a separate PR later.

**To be honest I am quite happy with current state of this PR, and I don't really see much that can be changed now.**

dhalbert reviewed (approved):
I tested this with the reactle game from https://learn.adafruit.com/wordle-personal-esp32-s2-web-server, on Firefox and Chrome. It worked fine.

@michalpokusa We really appreciate your work on this!

@anecdata Do you want to test further? I will merge after hearing if you if you think this is good to go.

I will release this as 1.0.0 since there are API changes.

anecdata reviewed (approved):
This looks really good. Tested with and without `content-length` client header.

Thanks, @michalpokusa!

michalpokusa commented:
Thank you! 😊
